rfcs/text/0011_global_search.md

```diff
@@ -0,0 +1,693 @@
+
+```ts
+coreStart.globalSearch.find('some term').subscribe(({ results, complete }) => {
+    showAsyncSearchIndicator(!complete);
+    updateResults(results);
+});
+```
+
+# Motivation
+
+Kibana should do it's best to assist users searching for and navigating to the various objects present on the Kibana platform.
+
+We should expose an API to make it possible for plugins to search for the various objects present on a Kibana instance.
+
+The first consumer of this API will be the global search bar [#57576](https://github.com/elastic/kibana/issues/57576). This API
+should still be generic to answer similar needs from any other consumer, either client or server side.
+
+# Detailed design
+
+## API Design
+
+### Result provider API
+
+#### common types
+
+```ts
+/**
+ * Static list of all the possible result types.
+ * Ordinal value matters here, as it will be used to sort results of different types.
+ */
+enum SEARCH_TYPE {
+  // non-exhaustive
+  applications = 10,
+  dashboard = 20,
+  visualization = 30,
+  search = 40,
+}
+
+/** @public */
+type GlobalSearchResultType = keyof typeof SEARCH_TYPE;
+
+/**
+ * Options provided to {@link GlobalSearchResultProvider | result providers} `find` method.
+ * Currently empty and only present to keep the API future-proof.
+ */
+interface GlobalSearchOptions {}
+
+/**
+ * Representation of a result returned by a {@linlkGlobalSearchResultProvider | result provider}
+ */
+interface GlobalSearchResult {
+  /** an id that should be unique for an individual provider's results */
+  id: string;
+  /** the title/label of the result */
+  title: string;
+  /** the type of result */
+  type: GlobalSearchResultType;
+  /** an optional EUI icon name to associate with the search result. If not specified, will use a default icon */
+  icon?: string;
+  /** The url to navigate to this result. This can be either an absolute url, or a path relative to the performing request's basePath */
+  url: string;
+  /** the score of the result, from 1 (lowest) to 100 (highest) */
+  score: number;
+  /** an optional record of metadata for this result */
+  meta?: Record<string, any>;
+}
+```
+
+#### server
+
+```ts
+/**
+ * Context passed to server-side {@GlobalSearchResultProvider | result provider}'s `find` method.
+ */
+export interface GlobalSearchProviderContext {
+  request: KibanaRequest;
+  core: {
+    savedObjects: {
+      client: SavedObjectsClientContract;
+      typeRegistry: ISavedObjectTypeRegistry;
+    };
+    elasticsearch: {
+      legacy: {
+        client: IScopedClusterClient;
+      };
+    };
+    uiSettings: {
+      client: IUiSettingsClient;
+    };
+  };
+}
+
+/**
+ * GlobalSearch result provider, to be registered using the {@link GlobalSearchSetup | global search API}
+ */
+type GlobalSearchResultProvider = {
+  id: string;
+  find(
+    term: string,
+    options: GlobalSearchOptions,
+    context: GlobalSearchProviderContext
+  ): Observable<GlobalSearchResult[]>;
```

Comment:
> We should include a preference option here in order to get consistent scoring from Elasticsearch 

Seems like a great addition. The `preference` term is really ES-specific, but as a majority (of not all expect the application one) of providers would be performing es query, I guess it's alright

> Core should pre-populate this preference with some sort of stable identifier for the user making the request to ensure that the results are consistent between keypresses.

Do you see that for the client-side, server-side, or both APIs? 

From the client-side, generating an arbitrary token that will live until a full reload seems easy (we can even persist it in local storage if we want a longer lifespan). 

From the server-side, not sure there is currently a way to get an identifier relative to a user 'session' to forge a stable user/session based identifier?

> I'm not sure if this value should be part of the options object or another argument. It may make sense to be a first-level argument in order to encourage implementations to use it properly.

I would prefer to keep it in the `options` TBH, just to avoid adding additional parameters to the signature when `options` is present for this exact purpose. Imho API documentation should be enough to encourage implementations to use it?

Another point: as we would be generating this token's default from the GS service when calling the underlying providers, should we let the API consumer manually defines this token if they want to, or should this only be an internal option between the service and the result providers hidden from the public API?

Reply:
> Do you see that for the client-side, server-side, or both APIs?
> 
> From the client-side, generating an arbitrary token that will live until a full reload seems easy (we can even persist it in local storage if we want a longer lifespan).
> 
> From the server-side, not sure there is currently a way to get an identifier relative to a user 'session' to forge a stable user/session based identifier?

On the server-side I believe security does provide a session identifier, however we can't rely on this since security may be disabled (and this is going in OSS code).

I don't see any problem with the client-side generating a UUID or similar. I think this is primarily a UX problem we want to solve for as-you-type experiences which does not apply to ChatOps. 

> Imho API documentation should be enough to encourage implementations to use it?

SGTM

> Another point: as we would be generating this token's default from the GS service when calling the underlying providers, should we let the API consumer manually defines this token if they want to, or should this only be an internal option between the service and the result providers hidden from the public API?

I lean towards making this API foolproof rather than more flexible (esp if we don't have a real usecase in mind). If we're generating this identifier on the client-side, maybe the easiest thing would be to expose the option on the HTTP API but not on the client-side JS API.

Reply:
Added in https://github.com/elastic/kibana/pull/64284/commits/cbdc44ce111f8353b77237dbeb6b0c7dd420f9ba